cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala

```diff
@@ -0,0 +1,212 @@
+    * @param request the report generation request
+    * @param inputURIs URIs of the avro files containing program run meta records
+    * @param reportIdDir location of the directory where the report files directory, COUNT file,
+    *                      and _SUCCESS file will be created.
+    * @throws java.io.IOException when fails to write to the COUNT or _SUCCESS file
+    */
+  @throws(classOf[IOException])
+  def generateReport(spark: SparkSession, request: ReportGenerationRequest, inputURIs: java.util.List[String],
+                     reportIdDir: Location): Unit = {
+    import spark.implicits._
+    val df = spark.read.format("com.databricks.spark.avro").load(inputURIs: _*)
+    // Get the fields to be included in the final report and additional fields required for filtering and sorting
+    val (reportFields: Set[String], additionalFields: Set[String]) = getReportAndAdditionalFields(request)
+    // Create an aggregator that aggregates grouped data into a column with data type Record.
+    val aggCol = new RecordAggregator().toColumn.alias(RECORD_COL).as[Record]
+    // TODO: configure partitions. The default number of partitions is 200
+    // Group the program run meta records by program run Id's and aggregate the grouped data with aggCol.
+    // The initial aggregated DataFrame will have two columns:
+    // With every unique field in reportFields and additionalFields, construct and add new columns from record column
+    // in aggregated DataFrame, in addition to the two initial columns "run" and "record"
+    val aggDf = (reportFields ++ additionalFields).foldLeft(df.groupBy(Constants.RUN).agg(aggCol))((df, fieldName) =>
+      df.withColumn(fieldName, df(RECORD_COL).getField(fieldName)))
+    // Filter the aggregated DataFrame
+    var resultDf = aggDf.filter(getFilter(request, aggDf))
+    // If sort is specified in the request, apply sorting to the result DataFrame
+    Option(request.getSort).foreach(_.foreach(sort => {
+      val sortField = aggDf(sort.getFieldName)
+      sort.getOrder match {
+        case Order.ASCENDING => {
+          resultDf = resultDf.sort(sortField.asc)
+          LOG.debug("Sort by {} in ascending order", sortField)
+        }
+        case Sort.Order.DESCENDING => {
+          resultDf = resultDf.sort(sortField.desc)
+          LOG.debug("Sort by {} in descending order", sortField)
+        }
+      }
+    }))
+    // drop the columns which should not be included in the report
+    resultDf.columns.foreach(col => if (!reportFields.contains(col)) resultDf = resultDf.drop(col))
+    resultDf.persist()
+    // Writing the DataFrame to JSON files requires a non-existing directory to write report files.
+    // Create a non-existing directory location with name ReportSparkHandler.REPORT_DIR
+    val reportDir = reportIdDir.append(ReportSparkHandler.REPORT_DIR).toURI.toString;
+    resultDf.coalesce(1).write.option("timestampFormat", "yyyy/MM/dd HH:mm:ss ZZ").json(reportDir)
```

Comment:
Why need to coalesce to 1?

Reply:
could this cause any memory issue?

Reply:
it is possible. Currently assume there is only one report file. Added a TODO in `ReportGenerationSpark` to handle more than one report files, and here will be improved together.

Reply:
Is there spark config to control the default?